### DISCUSSION TITLE
Drei oder vier Eingangskanäle,Three or four input channels

### DISCUSSION THREAD

**Discussion by michaelw84** [jamulussoftware/jamulus#1721] [Ideas]
Hallo,
zu erst ein großes Lob! Tolles Programm klappt super.

Jetzt mein Wunsch:
Es wäre super wenn es anstelle von zwei wählbaren Eingangskanälen, drei besser vier geben würde.
So hätte man die Möglichkeit ein Stereogerät und zusätzlich noch ein Mikrofon anzuschließen.

Hello,
first of all great praise!  Great program works great.

Now my wish:
It would be great if instead of two selectable input channels there were three, better four.
This would give you the option of connecting a stereo device and an additional microphone.

**Comment by ann0see** [comment]
First of all, welcome to the community!

Could you please tell us a bit more about your use case? (Keyboard + Mic, Guitar + Keyboard + mic,…?) maybe there’s an easy solution. 

I think your request is rather an advanced use case and not simple to implement with one client (we had multiple feature requests for that). However at least what I know from Linux, it’s quite simple to just run two clients. Windows should be possible too. 

You could also use a (virtual) mixer. 

Although I am not 100% sure, but with ASIO4ALL I could choose two inputs as one channel once. Didn’t try after I switched to JACK under Linux which made the whole process a lot easier.

**Reply by michaelw84** [reply]
Hi, thanks for the answer. For example, I like to use a Line6 Helix in stereo and a microphone. Our drummer uses stereo on his e-drums and a microphone. Our keyboardist uses two keyboards each mono and a microphone. Therefore, it is actually an extension that would benefit a broad mass of people.

I use a Windows computer and there the Asio driver for my interface is blocked as soon as it is used by an application. Therefore I cannot open a second instance and use the same interface. Some examples are Steinberg UR series, Focusrite Scarlett series, Presonus USB interface.

I've tried a lot with Jack, unfortunately the latency always leads below. The ASIO4All driver is significantly worse and slower in latency than the driver supplied for the interfaces.
We regularly play with the band on Jamulus and without Jack and without ASIO4All we have the best results in terms of latency.

Depending on the version, the interface already provides everything necessary since the various inputs, for example. can also be used in the DAWs.

It would be a significant improvement of Jamulus if three or four inputs of the same interface could be selected.

With best regards
Michael

**Reply by dcorson-ticino-com** [reply]
Your best be may be in inexpensive USB mixer.
Adding channels to the client would be a lot of work and the volunteers programming and maintaining Jamulus already have their plate pretty full.  I would not try to hold my breath.

**Reply by michaelw84** [reply]
Good morning, 
I didn't mean to be rude or demanding. That was just a suggestion from which I believe that many will benefit and that the program will be significantly upgraded. I'm in IT myself, but unfortunately not a programmer, otherwise I would have loved to invest my time in this feature. If no one has the inclination and time for this adaptation, 
I will respect that. If someone deals with it, I am of course happy.

**Reply by ann0see** [reply]
You weren’t rude in any way ;-)

If you’re an IT guy, you might want to look into JACK again, but this time on Linux. JACK on Linux + multiple clients works much nicer (and without adding too much latency) than anything on Windows or macOS (at least from my experience).

**Reply by jujudusud** [reply]
Hi @michaelw84 
In ordre to use JACK server with a very low latency on Windows 10 64 bits, you have to :
- install JACK 64 bits for Windows at the latest release available (1.9.18 for now),
- build and install a custom Jamulus with "jackonwindows" activated.

Only After that you will have the same latency and the capacity to interconnect whatever sources/playback/softwares you want.

We are currently trying to deliver a New autobuild version for custom jackonwindows but it is not yet finished.

I can help you understand how to do that.

**Reply by michaelw84** [reply]
Hi,
thank you for your support.
Unfortunately the Jack 64bit application always runs into an error: Could not connect to JACK server as client. - Overall operation failed. - Unable to connect to server. Please check the messages window for more info.

**Reply by jujudusud** [reply]
Ok, we have to investigate. Are you ok to search how to make this work ?
The first thing is that you have to update your Windows OS to the latest. Is it ok ?
After that download the latest JACK installer here : [JACK2 64 bits for Windows](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.18/jack2-win64-v1.9.18.exe)
Install. Is it ok ?
@henkdegroot could you please tell us how to run `Qjackctl` on windows and configure it properly ?

**Reply by henkdegroot** [reply]
> Unfortunately the Jack 64bit application always runs into an error: Could not connect to JACK server as client. - Overall operation failed. - Unable to connect to server.

This is most likely because the build is not for support with JACK (i.e. the standard jamulus build). As Jamulus will only be able to connect direct to JACK when using a version which has been compiled with the "jackonwindows" option.
If you want to test, you can download this version [Jamulus-3.8.0beta2-installer-win-withjack2](https://we.tl/t-JXoQQOcEnY).
Download link expires in two days....I can setup a new download link if need.

NOTE: This installer is for 64-bit windows only.

Please share the results!

**Reply by jujudusud** [reply]
for information it works here : 
![jackonwindows](https://user-images.githubusercontent.com/12748768/119178886-92faa200-ba6e-11eb-87e8-0cb5fcf95983.png)

**Reply by michaelw84** [reply]
Hello, 
please excuse the late response. I would love to test it, unfortunately, the link has expired, could you please set it again? 
I get the error message before I start jamulus. As soon as I start Jack and then click play, the error occurs. I've reinstalled Jack several times now. All installed software packages are up to date.

**Reply by henkdegroot** [reply]
@michaelw84 , you can download the windows installer with jack support [here](https://github.com/henkdegroot/jamulus/releases/tag/3.8.0rc1dev)
You may received a few warning during the install regarding missing ASIO, just click continue to complete the install.

You still need to have Jack installed and running first. The current version [JACK 1.9.18 win 64](https://github.com/jackaudio/jack2-releases/releases/download/v1.9.18/jack2-win64-v1.9.18.exe) can be downloaded here.
Use the default options when installing. After this is installed, you start the QJackCtl app. In the JACK Audio Connection Kit window, you need to click setup and ensure the Sample Rate is set to 48000 and Frames/Period to 128 or 64.
Once this is set you click the Start button to start the Jack Server.

Now you can start Jamulus and it should connect to the jack server.

Hope this helps.

**Reply by jujudusud** [reply]
The first step is to make JACK server run.
Use Qjackctl app and clic the setup button.
Ensuite that portaudio is used and realtime also.
Select your ASIO device in interface, 48kHz, and 128 frames/period.
If the ASIO doesn't appear just disconnect and reconnect your audio card (USB câble).
When that done, return to the main window and clic on start.
I will check where I Can put the last version of custom Jamulus to download.

**Reply by jujudusud** [reply]
"I use a Windows computer and there the Asio driver for my interface is blocked as soon as it is used by an application."
JACK server solves that problem. But to get it running you have to unlock your card. The safe way is disconnect reconnect USB câble.

**Reply by jujudusud** [reply]
"The ASIO4All driver is significantly worse and slower in latency than the driver supplied for the interfaces."
Please uninstall asio4all to avoid it's running.

**Reply by michaelw84** [reply]
> The first step is to make JACK server run.
> Use Qjackctl app and clic the setup button.
> Ensuite that portaudio is used and realtime also.
> Select your ASIO device in interface, 48kHz, and 128 frames/period.
> If the ASIO doesn't appear just disconnect and reconnect your audio card (USB câble).

Hi, 
I tested it exactly with the settings.
Unfortunately, when I start the server, I always get the error mentioned. I tried it with different audio interfaces.
Maybe Jack has a problem with another software or driver.

**Reply by jujudusud** [reply]
What is your ASIO device ?

**Reply by henkdegroot** [reply]
Also to be clear, please detail when you get the error. As you mention, when I start the server.....to me this means the Jack Server.
But when you start the Jack Server, there is no client connecting yet, so how can you get a message that you cannot connect to the Jack Server as the Client?

If this problem is when starting the Jack Server, I had issues with this as well it was caused by two things. One thing, the audio device I tried to use was already in use by another application and the other thing was timing related.

I have resolved the "timing" issue by increasing the "Start Delay" option in the "Advanced" tab of the settings window:
![qjackctl-advanced](https://user-images.githubusercontent.com/13550012/119475923-dfcfc880-bd4d-11eb-9088-ba4ae2772505.png)

Still no luck, even though this then seems to be a JACK issue and not a Jamulus issue, what messages do you see in the "message" window?
Here is what it looks like in mine after I start the Jack Server:
![qjackctl-messages](https://user-images.githubusercontent.com/13550012/119476183-1dccec80-bd4e-11eb-9075-8125ae5ec5c4.png)

**Reply by jujudusud** [reply]
https://jackaudio.org/faq/jack_on_windows.html

**Reply by jujudusud** [reply]
with reaper in jamulus :)
![reaper-jack-jamulus](https://user-images.githubusercontent.com/12748768/119865003-81f0db80-bf1b-11eb-9711-6b5dbc0a7ba9.png)

**Comment by npostavs** [comment]
The Windows and Mac backends support mixing multiple channels (ultimately, only a mono or stereo signal can be sent to the server), but only if the device presents exactly 4 input channels.

https://github.com/jamulussoftware/jamulus/blob/r3_7_0/windows/sound.cpp#L230-L231
https://github.com/jamulussoftware/jamulus/blob/r3_7_0/mac/sound.cpp#L701-L702

**Reply by michaelw84** [reply]
Hi, I didn't quite understand the four entrances. How can I use this with a USB interface, for example.
 greeting

**Reply by npostavs** [reply]
Probably you can't use it, because your USB interface doesn't present 4 input channels (even if it has 4 physical channels, it might present virtual channels as well. For example, my Motu M2 has 2 physical channels (that is, 2 sockets to plug in microphones), but additionally has an 4 virtual channels (2 loopback & 2 loopback mix), so Jamulus sees it as having 6 channels).

You could edit the source to ignore channels beyond the first 4 I suppose.

**Reply by michaelw84** [reply]
Hi,
Many thanks for your response and help.
